nova_plugin/server.py

```diff
@@ -1134,53 +1131,59 @@ def _get_keypair_name_by_id(nova_client, key_name):
+    attached_ports = set([interface.port_id for interface in interfaces])
+    attached_nets = set([interface.net_id for interface in interfaces])
 
-    server_id = get_openstack_id(ctx)
-    connected_ports = neutron_client.list_ports(device_id=server_id)['ports']
-
-    # not counting networks connected by a connected port since allegedly
-    # the connection should be on a separate port
-    connected_ports_networks = {port['network_id'] for port in
-                                connected_ports if port['id'] not in port_ids}
-    connected_ports_ids = {port['id'] for port in
-                           connected_ports}
-    disconnected_networks = [network_id for network_id in network_ids if
-                             network_id not in connected_ports_networks]
-    disconnected_ports = [port_id for port_id in port_ids if port_id not
-                          in connected_ports_ids]
-    if disconnected_networks or disconnected_ports:
+    already_attached_ports = [port_id for port_id in attached_ports
+                              if port_id in port_ids]
+    already_attached_nets = [net_id for net_id in attached_nets
+                             if net_id in network_ids]
+    if already_attached_ports or already_attached_nets:
         raise NonRecoverableError(
-            'Expected external resources to be connected to external server {'
-            '0}: Networks - {1}; Ports - {2}'.format(server_id,
-                                                     disconnected_networks,
-                                                     disconnected_ports))
+            'Several ports/networks already connected to external server '
+            '{0}: Networks - {1}; Ports - {2}'
+            .format(external_server.human_id,
+                    already_attached_nets,
+                    already_attached_ports))
+
+    # attach ports
+    for port_id in port_ids:
+        external_server.interface_attach(port_id=port_id, net_id=None,
+                                         fixed_ip=None)
+        ctx.logger.info(
+            'Successfully attached port {0} to device (server) id {1}.'
+            .format(port_id, external_server.human_id))
+
+    # check currently attached networks, can be attached by port
+    attached_nets = [interface.net_id
+                     for interface in external_server.interface_list()]
+    for net_id in network_ids:
+        if net_id not in attached_nets:
+            external_server.interface_attach(port_id=None, net_id=net_id,
+                                             fixed_ip=None)
+            ctx.logger.info(
+                'Successfully attached network {0} to device (server) id {1}.'
+                .format(net_id, external_server.human_id))
+        else:
+            ctx.logger.info(
+                'Skipping network {0} attachment, because it is already '
+                'attached to device (server) id {1}.'
+                .format(net_id, external_server.human_id))
```

Comment:
Should we check that port-ids and net-ids do not overlap?

Reply:
It will require make queries to openstack for get properties for ports, but we should ignore cases when ports proposed for connect is also connected to networks for connect. 

We will have such case always when:
```yaml
    relationships:
    - type: cloudify.relationships.contained_in
      target: private_network
    - type: cloudify.openstack.server_connected_to_port
      target: private_port
```
that we have in documentation examples, and private port from private network. And we can all ports before networks, and only after that check that all networks connects.